### PR TITLE
Revives a Cloud Fnc for BigQuery storage task

### PR DESCRIPTION
**Pull request by troyraen**
- [x]  Adds a Cloud Function that stores alert data to the BigQuery tables _alerts_ and _DIASource_, and emits a Pub/Sub stream
- [x] Adds relevant info to metadata tracking
- [x] Updates docs

Note that this cloud fnc is triggered from the main _alerts_ stream and not from the _alert_avros_ stream emitted by the Cloud Storage bucket. This is because loading to the _DIASource_ table requires that we extract/repackage some of the alert data. If we triggered from the bucket's stream, we'd have to download the file first.

I setup a (partial) **testing** instance that has been processing the live production _alerts_ stream since 9/5/21. You can see how it's working:
- [Cloud Function](https://console.cloud.google.com/functions/details/us-central1/ztf-store_in_BigQuery-storebq?project=ardent-cycling-243415) _store_in_BigQuery_
- [BigQuery tables](https://console.cloud.google.com/bigquery?project=ardent-cycling-243415&d=ztf_alerts_storebq&p=ardent-cycling-243415&t=alerts&page=table&ws=!1m10!1m4!4m3!1sardent-cycling-243415!2sztf_alerts_storebq!3salerts!1m4!4m3!1sardent-cycling-243415!2sztf_alerts_storebq!3sDIASource) _alerts_ and _DIASource_

@djperrefort will you review when you have a chance?
@wmwv FYI

### CODE REVIEW

File: broker/cloud_functions/store_BigQuery/main.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python3
+# -*- coding: UTF-8 -*-
+
+"""This module stores alert data in BigQuery tables."""
+
+import base64
+from google.cloud import logging
```
Comment (troyraen) [inline broker/cloud_functions/store_BigQuery/main.py:7]: I'm not sure how to fix this Codacy error, and the following one. Is there a requirements file/list somewhere that needs to be updated? If so, noting that there are more of these in #71.

Comment (djperrefort) [reply broker/cloud_functions/store_BigQuery/main.py:7]: Hmmm. This is a mypy issue. I'll take a look at it and let you know what I find.

File: broker/cloud_functions/store_BigQuery/main.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+bq_dataset = f"{SURVEY}_alerts"
+ps_topic = f"{SURVEY}-BigQuery"
+if TESTID != "False":
+    bq_dataset = f"{bq_dataset}_{TESTID}"
+    ps_topic = f"{ps_topic}-{TESTID}"
+
+schema_map = schema_maps.load_schema_map(SURVEY, TESTID)
+
+
+def run(msg: dict, context) -> None:
```
Comment (djperrefort) [inline broker/cloud_functions/store_BigQuery/main.py:31]: Missing type hint for `context`

File: broker/cloud_functions/store_BigQuery/main.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+        context: Metadata describing the Cloud Function's trigging event.
+    """
+    # extract the alert
+    alert_dict = data_utils.decode_alert(
+        base64.b64decode(msg["data"]), drop_cutouts=True, schema_map=schema_map
+    )
+
+    # send the alert to BigQuery tables
+    a_table = insert_rows_alerts(alert_dict)
```
Comment (djperrefort) [inline broker/cloud_functions/store_BigQuery/main.py:47]: What do `a` and `d` stand for in these variable names?

File: broker/cloud_functions/store_BigQuery/main.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    d_table = insert_rows_DIASource(alert_dict)
+
+    # announce what's been done
+    publish_pubsub(alert_dict, [a_table, d_table])
+
+
+def insert_rows_alerts(alert_dict):
```
Comment (djperrefort) [inline broker/cloud_functions/store_BigQuery/main.py:54]: Missing type hinting here and for all functions later on.

File: broker/cloud_functions/store_BigQuery/main.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    cand = {k: v for k, v in alert_dict["candidate"].items() if k not in dup_cols}
+
+    # get info for provenance
+    metakeys = ["schemavsn", "publisher", "objectId", "candid"]
+    metadict = {k: v for k, v in alert_dict.items() if k in metakeys}
+
+    # get string of previous candidates' candid, comma-separated
+    if alert_dict["prv_candidates"] is not None:
+        tmp = [pc["candid"] for pc in alert_dict["prv_candidates"]]
+        prv_candids = ",".join([f"{cid}" for cid in tmp if cid is not None])
```
Comment (djperrefort) [inline broker/cloud_functions/store_BigQuery/main.py:107]: This line is a double loop: once for the list constructor and another for the join. Remove the brackets to only loop once: `prv_candids = ",".join(str(cid) for cid in tmp if cid is not None)`

File: broker/cloud_functions/store_BigQuery/main.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    dup_cols = ["candid"]  # candid is repeated, drop the one nested here
+    cand = {k: v for k, v in alert_dict["candidate"].items() if k not in dup_cols}
+
+    # get info for provenance
+    metakeys = ["schemavsn", "publisher", "objectId", "candid"]
+    metadict = {k: v for k, v in alert_dict.items() if k in metakeys}
+
+    # get string of previous candidates' candid, comma-separated
+    if alert_dict["prv_candidates"] is not None:
+        tmp = [pc["candid"] for pc in alert_dict["prv_candidates"]]
```
Comment (djperrefort) [inline broker/cloud_functions/store_BigQuery/main.py:106]: Two (technically three) loops inside a conditional is more than my brain wants to think about. For the sake of being pythonic, why not:

```python
prv_candids = ",".join(str(pc["candid"]) for pc in alert_dict["prv_candidates"] if pc["candid"] is not None)
```

File: broker/cloud_functions/store_BigQuery/main.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python3
+# -*- coding: UTF-8 -*-
+
+"""This module stores alert data in BigQuery tables."""
+
+import base64
+from google.cloud import logging
+import os
+
+from broker_utils import data_utils, gcp_utils, schema_maps
```
Comment (djperrefort) [inline broker/cloud_functions/store_BigQuery/main.py:10]: I would just disable the missing stub analysis in codacy for now.

**Review by djperrefort — APPROVED** [review]
Some requested changes - all on small minor details.

File: broker/cloud_functions/store_BigQuery/main.py (hunk 6 above)
Comment (troyraen) [reply broker/cloud_functions/store_BigQuery/main.py:106]: Oh, excellent! Yes, thanks. I'm happy to write lines like this; I actually thought I wasn't supposed to let them run that long.

**Pull request merged**